src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java

```diff
@@ -0,0 +1,206 @@
+import software.amazon.awssdk.services.greengrassv2data.model.GetDeploymentConfigurationResponse;
+import software.amazon.awssdk.utils.IoUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.security.NoSuchAlgorithmException;
+import javax.inject.Inject;
+
+public class DeploymentDocumentDownloader {
+    private static final Logger logger = LogManager.getLogger(DeploymentDocumentDownloader.class);
+
+    private final GreengrassV2DataClient greengrassV2DataClient;
+    private final HttpClientProvider httpClientProvider;
+
+    private String thingName;   // not final as it could be dynamically changed
+
+
+    /**
+     * Constructor.
+     *
+     * @param greengrassServiceClientFactory greengrassServiceClientFactory
+     * @param deviceConfiguration            deviceConfiguration
+     * @param httpClientProvider             httpClientProvider
+     */
+    @Inject
+    public DeploymentDocumentDownloader(GreengrassServiceClientFactory greengrassServiceClientFactory,
+                                        DeviceConfiguration deviceConfiguration,
+                                        HttpClientProvider httpClientProvider) {
+        this.greengrassV2DataClient = greengrassServiceClientFactory.getGreengrassV2DataClient();
+        deviceConfiguration.getThingName().subscribe((why, newv) -> thingName = Coerce.toString(newv));
+        this.httpClientProvider = httpClientProvider;
+    }
+
+    /**
+     * Download the full deployment document stored from greengrass cloud.
+     *
+     * @param deploymentId deployment id
+     * @return DeploymentDocument
+     * @throws DeploymentDocumentDownloadException if failed to download the full deployment document.
+     */
+    public DeploymentDocument download(String deploymentId) throws DeploymentDocumentDownloadException {
+        // TODO [Next Revision] Support downloading retrials for intermittent network and temporary GG data plane
+        //  service error with RetryUtils
+
+        // 1. Get url, digest, and algorithm by calling gg data plane
+        GetDeploymentConfigurationResponse response = getDeploymentConfiguration(deploymentId);
+
+        String preSignedUrl = response.preSignedUrl();
+        String algorithm = response.integrityCheck().algorithm();
+        String digest = response.integrityCheck().digest();
+
+        validate(preSignedUrl, algorithm, digest);
+
+        // 2. Download configuration and check integrity
+        String configurationInString = downloadFromUrl(deploymentId, preSignedUrl);
+        checkIntegrity(digest, configurationInString);
+
+        // 3. deserialize and convert to device model
+        Configuration configuration = deserializeDeploymentDoc(configurationInString);
+        return DeploymentDocumentConverter.convertFromDeploymentConfiguration(configuration);
+
+    }
+
+    private String downloadFromUrl(String deploymentId, String preSignedUrl)
```

Comment:
retries?

Reply:
I left it in next revision. Need to think a little bit more on the retry strategy